src/main/java/service/impl/ShopServiceImpl.java

```diff
@@ -0,0 +1,39 @@
+package service.impl;
+
+import db.Storage;
+import java.util.List;
+import model.FruitTransaction;
+import service.ShopService;
+import strategy.OperationStrategy;
+
+public class ShopServiceImpl implements ShopService {
+    private OperationStrategy operationStrategy;
+
+    public ShopServiceImpl(OperationStrategy operationStrategy) {
+        this.operationStrategy = operationStrategy;
+    }
+
+    @Override
+    public void process(List<FruitTransaction> transactions) {
+        for (FruitTransaction transaction : transactions) {
+            if (!Storage.reports.containsKey(transaction.getFruit())) {
+                Storage.reports.put(transaction.getFruit(), getHowManyHaveChanged(transaction));
+            } else {
+                Storage.reports.put(transaction.getFruit(), getBalance(transaction)
+                        + getHowManyHaveChanged(transaction));
+            }
+            if (getBalance(transaction) < 0) {
+                throw new RuntimeException("The quantity is not enough to purchase");
+            }
+        }
+    }
+
+    private int getHowManyHaveChanged(FruitTransaction transaction) {
+        return operationStrategy.getOperationHandler(transaction.getOperation())
+                .apply(transaction.getQuantity());
+    }
```

Comment:
It's better to handle the case where `operationStrategy.getOperationHandler(transaction.getOperation())` returns `null`, indicating that there's no operation handler for a given transaction type.